### DISCUSSION TITLE
API Fetch Combination of Pagination and Search Function

### DISCUSSION THREAD

**Discussion by MonaHer** [neuefische/questions#642] [Web]
Hi,
so I think my question is not really about a bug in my code, but it's more a logical question, I guess :).
So I fetched my data via an API and implemented a pagination:  In sum the fetched data is 1800 objects and not to fetch all the data all the time, with the pagination there are 20 objects per fetch. This worked all fine. Now I implemented a search function, that should search through all the data of the API. Now I have the problem, that 
only these two scenarios do work:

1. The pagination still works, but the search function only goes through the 20 fetched objects
or 
2. I fetch all the data, all data is displayed, the search function goes through all data, but then actually the pagination is redundant, no?

So I was trying and trying and trying to fetch the data, so that is displayed only paginated, but that the search function goes through all the data. So my question is: Is this even possible without backend? I've read something, that there is a way to do this without backend, but I was really not able to implement it into my code. 

Thank you so much in advance!

**Comment by Roland-Hufnagel** [comment]
Hi MonaHer, 
do you fetch the data from your own API  or from an external API? Most APIs offer a filter function just by adding a search string to the API route like 'https://rickandmortyapi.com/api/character/?name=rick&status=alive'. Then you'll recieve the filtered data incl. pagination. 

(If it is an external API it will work without backend)

Hope this can help, if not, please offer more information.

**Comment by MonaHer** [comment]
Hi Roland,
thank you for your answer! :) 
I do fetch from an external API.
This is my url: 
`https://api.smk.dk/api/v1/art/search/?keys=*&fields=image_thumbnail&fields=titles&fields=id&fields=production&fields=dimensions&fields=current_location_name&fields=production_dates_notes&fields=labels&filters=[image_hq:true],[object_names:painting],[public_domain:true]&offset=${
        offset - rowsPerPage
      }&rows=${rowsPerPage}&lang=en`

Your answer might help me a lot, if I got you right: So do you mean, I could implement a variable for example name=${name} (if I want to search by a specific name) into the url string and connect this variable with the input field of the search function? So if I would write for example "Matisse" into the search field input this string would become the value of ${name}?

**Reply by MonaHer** [reply]
Hello :)
I'm still struggling with that one. 
My fetch is in the app.js, here's the current status on my branch feature/navigation: [Link to fetch in my repo](https://github.com/MonaHer/capstone-mona/blob/main/pages/_app.js). 

I followed Roland's suggestion and added a search string in the url: `[title:${searchTerm}]` this works fine, but I still have the problem, that 
1. the fetch is paginated, but so the search function doesn't go through all data
or
2. the fetch is not paginated, but so the search function goes through all data

This depends on the value of `rowsPerPage` of the "not mutated" fetch. I think the value needs to be the same as of the mutated fetches, only then the functions of the mutated fetches do work. 

How can I solve this problem? :)

Thank you very much in advance!

**Comment by ahohnsen** [comment]
Hey Mona,
As @Roland-Hufnagel already mentioned, you can use the filter function provided by the external API. The SMK API provides a specific endpoint for the search: [/art/search](https://api.smk.dk/api/v1/docs/#/Artworks/searchArt)

Check out the link! 👀 You can try it out there before implementing it in your app. You can also specify to which fields you want to limit the search. 💃🏼

**Reply by MonaHer** [reply]
Hi Andrea, 
thank you for your reply! :) I still don't understand how this could fix my bug, as I already fetch all my data via the art/search endpoint.

**Reply by ahohnsen** [reply]
According to the documentation of the API you fetch all artwork pieces with `https://api.smk.dk/api/v1/art/search/?keys=*`, since you are searching for `*`. You need to replace the `*` with the query string from your search input field. 
An example:

If your user is searching for "monk", the URL you need to fetch data from will be: `https://api.smk.dk/api/v1/art/search?keys=monk`

In this way, you will not search the data you have already fetched. Every time the user is using your search bar, your app will fetch data from your API, but only the data matching the query string.
